lua/luasnip/extras/_treesitter.lua

```diff
@@ -0,0 +1,476 @@
+	assert(#lines == 1)
+	local current_line = lines[1]
+
+	local o = {
+		ori_bufnr = ori_bufnr,
+		ori_line = current_line,
+		ori_row = row,
+		ori_col = col,
+		match = match,
+	}
+	setmetatable(o, {
+		__index = FixBufferContext,
+	})
+
+	return o
+end
+
+function FixBufferContext:enter()
+	local current_line_left = self.ori_line:sub(1, self.ori_col - #self.match)
+	local current_line_right = self.ori_line:sub(self.ori_col + 1)
+
+	vim.api.nvim_buf_set_lines(
+		self.ori_bufnr,
+		self.ori_row - 1,
+		self.ori_row,
+		true,
+		{ current_line_left .. current_line_right }
+	)
+	local parser, source =
+		vim.treesitter.get_parser(self.ori_bufnr), self.ori_bufnr
+	parser:parse()
+
+	return parser, source
+end
+
+function FixBufferContext:leave()
+	vim.api.nvim_buf_set_lines(
+		self.ori_bufnr,
+		self.ori_row - 1,
+		self.ori_row,
+		true,
+		{
+			self.ori_line,
+		}
+	)
+	local parser, source =
+		vim.treesitter.get_parser(self.ori_bufnr), self.ori_bufnr
+	parser:parse()
+	return parser, source
+end
+
+local function wrap_enter_and_leave_func(ori_bufnr, match)
+	local context = FixBufferContext.new(ori_bufnr, match)
+
+	return function()
+		context:enter()
+	end, function()
+		context:leave()
+	end
+end
+
+local builtin_tsnode_selectors = {
+	any = function(captures)
+		local best_match
+		return {
+			match = function(capture_name, node)
+				if captures and not captures[capture_name] then
+					return false
+				end
+				if best_match == nil then
+					best_match = node
+				end
+				return true
+			end,
+			get_best_match = function()
+				return best_match
+			end,
+		}
+	end,
+	shortest = function(captures)
+		local best_match
+		local best_match_start
+		return {
+			match = function(capture_name, node)
+				if captures and not captures[capture_name] then
+					return false
+				end
+
+				local start_row, start_col, _, _ =
+					vim.treesitter.get_node_range(node)
+				if
+					(best_match == nil)
+					or (start_row > best_match_start[1])
+					or (
+						start_row == best_match_start[1]
+						and start_col > best_match_start[2]
+					)
+				then
+					best_match = node
+					best_match_start = { start_row, start_col }
+				end
+				return true
+			end,
+			get_best_match = function()
+				return best_match
+			end,
+		}
+	end,
+	longest = function(captures)
+		local best_match
+		local best_match_start
+		return {
+			match = function(capture_name, node)
+				if captures and not captures[capture_name] then
+					return false
+				end
+
+				local start_row, start_col, _, _ =
+					vim.treesitter.get_node_range(node)
+				if
+					(best_match == nil)
+					or (start_row < best_match_start[1])
+					or (
+						start_row == best_match_start[1]
+						and start_col < best_match_start[2]
+					)
+				then
+					best_match = node
+					best_match_start = { start_row, start_col }
+				end
+				return true
+			end,
+			get_best_match = function()
+				return best_match
+			end,
+		}
+	end,
```

Comment:
WDYT about changing the selectors so they can abort the match-iteration? Seems like a worthwhile optimization, considering `any` would iterate all matches, when the first match would do.

Reply:
But we also want to get all matches and set these into `env`.

Reply:
I don't think we need all matches.. what would that be useful for?
What I think would be useful to collect, and provide to the user, are other captures of the one match we choose, that way it's easy to extract more information with a custom-query (for example, trigger behind any function-call, but also capture the name of the function, and make that accessible)

Reply:
Sometimes we want not only the best match which will be a part of expansion, but also others matches which make up an expansion environment. For example, `test.fn` will expand to `fn test(&self) {}` in the `impl` block, and expand to `fn test() {}` else.

Reply:
Mhmmm, I think `conditon` should be used for stuff like that..
Do you have an example where you use this?

Reply:
It can be impl in multiple snippets with conditions. But I will cause `resolveExpandParams` which reparse the buffer run multiple times. An example like this:

```lua
  tsp.treesitter_postfix({
    trig = ".fn",
    name = "void xxx()",
    dscr = "Completes a function declaration",
    matchTSNode = {
      query = [[
      [
        (call_expression)
        (identifier)
        (template_function)
        (subscript_expression)
        (field_expression)
        (user_defined_literal)
      ] @any_expr
      [
        (argument_list)
        (parameter_list)
      ] @in_arg_list
      [
        (function_definition)
        (lambda_expression)
        (field_declaration)
      ] @in_func_body
      [
        (struct_specifier)
        (class_specifier)
      ] @in_class_body
      ]],
      select = {},
    },
  }, {
    d(1, function(_, parent)
      local env = parent.env
      local last_type, last_type_row, last_type_col
      local expand_env = {}
      expand_env["CPP_IN_ARGUMENT"] =
        find_nearest_capture(env.TREESITTER_MATCHES, "in_arg_list")
      expand_env["CPP_IN_FUNCTION_BODY"] =
        find_nearest_capture(env.TREESITTER_MATCHES, "in_func_body")
      expand_env["CPP_IN_CLASS_BODY"] =
        find_nearest_capture(env.TREESITTER_MATCHES, "in_class_body")

      local keys = {
        "CPP_IN_ARGUMENT",
        "CPP_IN_FUNCTION_BODY",
        "CPP_IN_CLASS_BODY",
      }
      for _, key in ipairs(keys) do
        if env[key] ~= nil then
          if last_type == nil then
            last_type = key
            last_type_row = env[key][1]
            last_type_col = env[key][2]
          else
            if
              last_type_row < env[key][1]
              or (last_type_row == env[key][1] and last_type_col < env[key][2])
            then
              last_type = key
              last_type_row = env[key][1]
              last_type_col = env[key][2]
            end
          end
        end
      end

      if
        last_type == "CPP_IN_ARGUMENT" or last_type == "CPP_IN_FUNCTION_BODY"
      then
        return lambda_snippet_node
      else
        return make_function_snippet_node(env)
      end
    end, {}),
  }),
```

Reply:
Add a new builtin `any_all` and make `any` short circuit.

Reply:
> But I will cause resolveExpandParams which reparse the buffer run multiple times

One could also keep it in one snippet, and do the query for which case should actually be activated in the dynamicNode, and return the corresponding snippet.
This may actually be less work, since 
* the postfix really only needs to match a node in front of the trigger, so we could (and should, for better performance in the generic case) limit the buffer-region which is matched by the query to only the buffer before the trigger, which makes this "return all matches" not really work (nodes that contain the current node may end far behind it, and would (I'm not sure about this) not be included in the matches)
* I think your constraint would be more efficiently implemented not by a query, but by a simple walk, for example with `find_parent`

Besides that, returning all matches seems unexpected. I think we should return exactly one match with all its captures